### PR TITLE
Merge 0.7 branch (rebased to master)

### PR DESCRIPTION
**Pull request by ranjithshegde**
Since the 0.7 branch had no one taking initiative, this is essentially #1838  rebased to master.

I have been running the branch of the PR since it was created,  have zero issues. 
Seeing as 0.7.2 is released and master already requires nvim >=0.7, it seems like the ideal time to revisit and merge this.

All credits go to the excellent work of mjlbach, who I wont tag to respect his privacy.

### CODE REVIEW

**Comment by justinmk** [comment]
> I have been running the branch of the PR since it was created, have zero issues.

Thank you! That is valuable to know.

Why again was `commands` renamed to `user_commands` ?

**Comment by ranjithshegde** [comment]
> > I have been running the branch of the PR since it was created, have zero issues.
> 
> Thank you! That is valuable to know.
> 
> Why again was `commands` renamed to `user_commands` ?

This is straight from the comments on the main PR

"rename commands -> user commands to match API and free up commands
namespace"

On the one hand, the api resemblance is nice. But do you think its too confusing for the users? as these commands are essentially wrappers around unique `lsp/commands`?

**Comment by justinmk** [comment]
"API resemblance" here means "it literally matches `nvim_create_user_command` ? This is questionable because:

1. `nvim_create_user_command` itself is probably over-specific
2. this is not a strong enough motivation to ask users to perform a migration

**Comment by ranjithshegde** [comment]
> "API resemblance" here means "it literally matches `nvim_create_user_command` ? This is questionable because:
> 
> 1. `nvim_create_user_command` itself is probably over-specific and should have been named `nvim_create_command` (if ever it becomes theoretically possible to define non-user commands, why would we want a separate interface?! and, if it never becomes possible, what's the point of mentioning it over and over? Not even `:command` itself mentions this, it's literaly named `:command`.)
> 2. this is not a strong enough motivation to ask users to perform a migration

Will change it back to command

**Comment by justinmk** [comment]
Looking closer, and reviewing https://github.com/neovim/nvim-lspconfig/pull/1838 ,  I see that a migration will be required anyway, because the command structure changed to nvim_create_user_command. Matching the target structure is a good choice, but I wonder if instead we should maintain the old interface. And deprecate it. Because it's not really needed, people can just call `nvim_create_user_comand` from `on_init` or `on_attach`.

**Comment by ranjithshegde** [comment]
> Looking closer, and reviewing #1838 , I see that a migration will be required anyway, because the command structure changed.
> 
> I wonder if instead we should maintain the old interface. And deprecate it. Because it's not really needed, people can just call `nvim_create_user_comand` from `on_init` or `on_attach`.

Perhaps we remove the interface for user to define the commands, seeing as they could just use `nvim_create_user_command` where appropriate. But many internally defined commands (`ClangdSwitchSourceHeader` as an example) rely on this mechanism.
Not sure if any language specific plugins make use of them either.

So perhaps just remove it from user_space?

**Comment by justinmk** [comment]
> Perhaps we remove the interface for user to define the commands, seeing as they could just use `nvim_create_user_command` where appropriate.

I'm in favor of that.

> EDIT: Defining the commands in the setup/config of the language_server has the advantage of being able to access the server specific functions with little code.
>
> An example from the help.txt of this PR

I don't see the advantage. The `nvim_create_user_command` call could just reference `require'lspconfig'.texlab.buf_build`.

**Comment by ranjithshegde** [comment]
> I don't see the advantage. The `nvim_create_user_command` call could just reference `require'lspconfig'.texlab.buf_build`.

 Many methods from this repo (chiefly `lspconfig/utils` like `vim.fs`) are already in core, along with the edition of au-events like `LspAttach`. 
Maybe after the release of nvim 0.8 +  an appropriate "settle-in" period, this repo could be stripped down to a growing collection of lsp-configs with strong defaults?

With that in mind, this PR could  supply a deprecation warning concerning `commands`, with some examples of how to set the `user_commands` in an `on_attach` function?
Removing it now might be too radical for some users just yet!

**Comment by ranjithshegde** [comment]
I have added a deprecation notice for commands as discussed. There is also an example in `lspconfig.txt` to set them up in `on_attach` function.

Still not entirely sure this is ideal, having commands in individual LSP setups seems cleaner than having them in a common `on_attach` function where the commands are filtered by `client.name`. 

@justinmk Would appreciate your review and thoughts on this matter

**Comment by ranjithshegde** [comment]
@justinmk  Gentle ping
Requesting your review

**Comment by ranjithshegde** [comment]
Not sure who the maintainers are now to request review from.
@williamboman?

File: plugin/lspconfig.lua
Comment (kylo252) [inline plugin/lspconfig.lua:55]:
```suggestion
  local server_name = info.args and info.args ~= ""
  if server_name then
    if configs[server_name] then
      configs[server_name].launch()
    end
  else
    local other_matching_configs = require('lspconfig.util').get_other_matching_providers(vim.bo.filetype)
    for _, config in ipairs(other_matching_configs) do
      config.launch()
    end
  end
```

there seems to be a bug with using `fargs` and `nargs='?'`

File: test/minimal_init.lua
Comment (kylo252) [inline test/minimal_init.lua:34]:
```suggestion
```

should also remove the complimentary `end`

File: lua/lspconfig/server_configurations/zk.lua
Comment (williamboman) [reply lua/lspconfig/server_configurations/zk.lua:9]: Feels like such a breaking removal would belong in a separate PR?

File: lua/lspconfig/configs.lua
Comment (williamboman) [inline lua/lspconfig/configs.lua:98]:
```suggestion
          group = lsp_group,
          desc = string.format("Checks whether server %s should attach to a newly opened buffer inside workspace %q.", config.name, root_dir)
```

File: plugin/lspconfig.lua
Comment (williamboman) [reply plugin/lspconfig.lua:55]: Doesn't that suggestion turn ´server_name` into a boolean (in the truthy case, where it matters)?

File: plugin/lspconfig.lua
Comment (williamboman) [inline plugin/lspconfig.lua:2]: You never know when 1.0 is out 🤡 

```suggestion
if version_info.major < 0 and version_info.minor < 7 then
```

File: scripts/docgen.lua
Comment (williamboman) [inline scripts/docgen.lua:114]: What is there to be fixed?

File: lua/lspconfig/configs.lua
Comment (williamboman) [reply lua/lspconfig/configs.lua:284]: I fully agree with this. Very disruptive change that is difficult to motivate.

File: lua/lspconfig/configs.lua
Comment (williamboman) [inline lua/lspconfig/configs.lua:74]:
```suggestion
        group = lsp_group,
        desc = "Checks whether server %s should either start a new instance or attach to an existing one."
```

File: plugin/lspconfig.lua
Comment (williamboman) [inline plugin/lspconfig.lua:61]: This style of completion (uses `:command-completion-customlist` under the hood) has a worse autocompletion experience than the previous `:command-completion-custom`. I think we should stick with the former behaviour, and do something like:

```suggestion
  complete = "custom,v:lua.lsp_complete_configured_servers"
```

Comment (williamboman) [reply plugin/lspconfig.lua:61]: (for this we'll need to restore the `lsp_complete_configured_servers` global)

File: plugin/lspconfig.lua
Comment (williamboman) [inline plugin/lspconfig.lua:73]:
```suggestion
  complete = "custom,v:lua.lsp_get_active_client_ids"
```

File: plugin/lspconfig.lua
Comment (williamboman) [inline plugin/lspconfig.lua:83]:
```suggestion
  complete = "custom,v:lua.lsp_get_active_client_ids"
```

**Review by williamboman — COMMENTED** [review]
I'm personally really opposed to breaking the command API, if the plan is to deprecate it anyway.

File: lua/lspconfig/configs.lua
Comment (ranjithshegde) [reply lua/lspconfig/configs.lua:284]: That is a long story. In the original PR, mjlbach wanted create the entire interface for defining per-language-server commands called `user_commands` with identical signature to `nvim_create_user_comman`.

When I opened this PR (still 95% his work), other maintainers first suggested it be just called `commands` instead, which I changed, then it was discussed it was best to deprecate it entirely.

Right now it reflect all of those changes, in a single PR!. What is the suggestion? Keep it as the original PR by mjlbach and then deprecate it in the next nvim release cycle? Or something else?

File: lua/lspconfig/server_configurations/zk.lua
Comment (ranjithshegde) [reply lua/lspconfig/server_configurations/zk.lua:9]: Is the deprecation still being discussed or has it been already decided?

I am neither a maintainer nor a significant contributor  so I dont know if my opinion counts for anything, but for what its wroth:
I guess for certain servers features (like `clangd/switchSourceHeader`) these are considered as "extended" features, and certainly off 3.16 spec. 
But for servers like `texlab`, where the buffer holds more than 80% plain text and very few lines of latex commands, `texlab/build` is basically a primary feature. 

Without that command internally defined, users will have to resort to defining `params`, making the lsp call, extracting the results contextually and displaying them,  and then writing a `user_command` to call it. Which most users may not be able to do. Kind of removes the quick out of quickstart feature of the plugin. 

Unless someone writes a plugin for just `texlab/build` and `texlab/forward` commands. And such plugins for each server that has those command features.

Of course, if that is the consensus I will remove/or deprecate it!

File: plugin/lspconfig.lua
Comment (kylo252) [reply plugin/lspconfig.lua:55]: > Doesn't that suggestion turn ´server_name` into a boolean (in the truthy case, where it matters)?

it's specifically that `<f-args>` will get a string when using `nargs=?`, so `fargs[1]` doesn't really do anything useful, since calling `LspStart` without any arguments will give `server_name=""`

Comment (williamboman) [reply plugin/lspconfig.lua:55]: I meant won't the following expression

```lua
local server_name = info.args and info.args ~= ""
```

always evaluate to a boolean? It's later used to index the `configs` table: `configs[server_name]`.

I'd probably do something like the following:

```lua
if info.args ~= "" then
  local server_name = info.args
  if configs[server_name] then
    configs[server_name].launch()
  end
else
  -- ...
end
```

Comment (kylo252) [reply plugin/lspconfig.lua:55]: dang 🤦 

I meant to write it like this

```lua
local server_name = info.args ~= "" and info.args
```

but yeah, just need to make sure we're checking that args isn't an empty string

File: plugin/lspconfig.lua
Comment (justinmk) [reply plugin/lspconfig.lua:2]: `vim.fn.has('nvim-0.8')` is the correct way to check this

File: doc/lspconfig.txt
Comment (justinmk) [inline doc/lspconfig.txt:261]: why was it changed from a map?

File: lua/lspconfig/configs.lua
Comment (justinmk) [inline lua/lspconfig/configs.lua:43]: - soft deprecation is correct. Only show a warning/error, but continue to function
- as others mentioned, don't change the structure. we don't want a breaking change yet.

**Comment by justinmk** [comment]
> When I opened this PR (still 95% his work), other maintainers first suggested it be just called commands instead, which I changed, then it was discussed it was best to deprecate it entirely.
>
> Right now it reflect all of those changes, in a single PR!. What is the suggestion? Keep it as the original PR by mjlbach and then deprecate it in the next nvim release cycle? Or something else?

@ranjithshegde this is my fault, sorry. Let's see if we can make progress here without too much work. Here's what I suggest:

1.  soft deprecate (show warning/error but continue to work) the `commands` API without a breaking change (so don't need to change the code)
2. add the documentation examples for `nvim_create_user_command`
3. remove the examples for `commands`

File: doc/lspconfig.txt
Comment (ranjithshegde) [reply doc/lspconfig.txt:261]: As I understand it, this is to make the interface identical to `nvim_create_user_command`

File: doc/lspconfig.txt
Comment (ranjithshegde) [reply doc/lspconfig.txt:275]: in `lspconfig/server_configurations/texlab.lua` it does not accept a `range` key

File: lua/lspconfig/server_configurations/zk.lua
Comment (williamboman) [reply lua/lspconfig/server_configurations/zk.lua:9]: > Is the deprecation still being discussed or has it been already decided?
>
> I am neither a maintainer nor a significant contributor so I dont know if my opinion counts for anything, but for what its wroth:

Likewise. If by deprecation you mean these commands for `zk` then no, I don't believe it has been discussed before (quick GitHub search).

My take on this is: I really don't see the point of removing commands just for the sake of nudging users to install a dedicated plugin (https://github.com/mickael-menu/zk-nvim) for it (even though that plugin looks very thorough and well-made). It just adds unnecessary friction. If there are people that rely on these commands it's obviously providing value to them, and I think that weighs heavier than how to technically slice and dice things. If the commands ever stop working, or if they no longer suffice, these users have a very strong personal incentive to switch to https://github.com/mickael-menu/zk-nvim themselves.

Comment (ranjithshegde) [reply lua/lspconfig/server_configurations/zk.lua:9]: My suggestion would be to keep the internals, which are defined in `server_configurations/[lsp].lua` but deprecate the interface for defining them in setup as @justinmk  recommended. 

It is very trivial to just define custom commands in `nvim_create_user_command` instead of having`lspconfig` scan the the configs and define them using the same interface.

Comment (williamboman) [reply lua/lspconfig/server_configurations/zk.lua:9]: Sounds extremely reasonable to me!

File: lua/lspconfig/configs.lua
Comment (ranjithshegde) [reply lua/lspconfig/configs.lua:293]: What is the consensus on this? Should I revert it back to as it was?

Comment (ranjithshegde) [reply lua/lspconfig/configs.lua:293]: Thanks! I was in the process of writing something similar. 
Will have something ready for review by (my) today evening.

Comment (williamboman) [reply lua/lspconfig/configs.lua:293]: I updated the following part:

```
         local option, value = unpack(vim.split(v, "=", { plain = true }))
        option = option.gsub(v, "^%-+", "")
        opts[opts_aliases[option] or option] = value or true
```

This function is worthy of a test case or two (or five).

Comment (ranjithshegde) [reply lua/lspconfig/configs.lua:293]: The command table is already being validated in `configs.lua`
so simply this will do
```lua
function M.create_module_commands(module_name, commands)
  for command_name, def in pairs(commands) do
    local opts = {}
    if def.description and def['description'] ~= '' then
      opts.desc = def.description
    end
    api.nvim_create_user_command(command_name, function(...)
      require('lspconfig')[module_name].commands[command_name][1](...)
    end, opts)
  end
end
```
I am not even sure if the `description` key needs to be verified as its also being explicitly validated

Comment (williamboman) [reply lua/lspconfig/configs.lua:293]: Wont that suddenly exclude options like:

```lua
MyCommand = {
  function () end,
  ["-nargs=?"],
  ["-complete=..."],
  ["-bang"] = true
}
```

Comment (ranjithshegde) [reply lua/lspconfig/configs.lua:293]: I did check through all the default configurations, and they all only provide 
```lua
name = {
function() stuff end,
description = "whatever"
}
```
Let me see what the docs specify, or if those options have been allowed in the past

File: scripts/docgen.lua
Comment (ranjithshegde) [reply scripts/docgen.lua:114]: I think this has to be reverted as well

File: lua/lspconfig/configs.lua
Comment (williamboman) [inline lua/lspconfig/configs.lua:43]: While I think this deprecation message makes a lot of sense to have, it's pretty disruptive at the moment. Especially considering that some of the provided server configurations still make use of this API (i.e. nothing a user can fix themselves). May I suggest breaking the deprecation down in steps?

1) Soft-soft deprecation (deprecation notices in documentation and release notes only). This will allow 3rd party plugins time to move away from it (rust-tools.nvim is a pretty popular one, as an example).
2) Soft deprecation via warning message.
3) Removal of the feature.

**Review by williamboman — APPROVED** [review]
Awesome work! Code LGTM, just one comment regarding the deprecation message that I think is worth revisiting. I'll run this branch going forward and will report back if I run into issues

File: lua/lspconfig/configs.lua
Comment (ranjithshegde) [reply lua/lspconfig/configs.lua]: Removed the notification. I have added something in the commit message + in `lspconfig.txt`

Also, I have been running this branch from when it was started by mjlbach. Have had no issues. 
In fact, it improves a lot of usage options as `nvim_exec_autocmd` for `lspconfig` can now be called if you are defining/configuring a server not on `FileType` event.

File: plugin/lspconfig.lua
Comment (ii14) [inline plugin/lspconfig.lua:62]: This can be a lua function, no need to create global variables. I don't remember if this was supported right from the start, or was implement in later neovim versions though.
```suggestion
  complete = lsp_complete_configured_servers,
```

File: plugin/lspconfig.lua
Comment (ii14) [inline plugin/lspconfig.lua:1]: Add a guard here, so it can be blocked from loading.
```suggestion
if vim.g.lspconfig ~= nil then
  return
end
vim.g.lspconfig = 1

local version_info = vim.version()
```

File: plugin/lspconfig.lua
Comment (ii14) [inline plugin/lspconfig.lua:13]: Inconsistent use, sometimes these local variables are used, sometimes `require` is called again. In my opinion you should inline these and always call `require`, to not unnecessarily force these modules to be loaded.

File: plugin/lspconfig.lua
Comment (ranjithshegde) [reply plugin/lspconfig.lua:62]: This was the initial state of the PR. @williamboman requested the use of `custom, v:lua` method

File: plugin/lspconfig.lua
Comment (ii14) [reply plugin/lspconfig.lua]: @williamboman worse experience? What do you mean? It should be the same thing, except that `custom` should just expect a newline-separated string, and `customlist` a list.

File: plugin/lspconfig.lua
Comment (ranjithshegde) [reply plugin/lspconfig.lua:13]: `configs` is used several times in the file. Perhaps I only inline `utils`?

File: plugin/lspconfig.lua
Comment (ii14) [reply plugin/lspconfig.lua:62]: Okay, I think the issue was that you didn't filter completions, like this:
```lua
function complete(arg)
  local res = {}
  for _, entry in ipairs(lsp_complete_configured_servers()) do
    if entry:sub(1, #arg) == arg then
      table.insert(res, entry)
    end
  end
  return res
end
```
This can happen inside `lsp_complete_configured_servers` of course.

File: plugin/lspconfig.lua
Comment (ii14) [reply plugin/lspconfig.lua:13]: It's just three times basically.

File: plugin/lspconfig.lua
Comment (ranjithshegde) [reply plugin/lspconfig.lua:62]: The original functions were these
```lua
local lsp_get_active_client_ids = function()
  return vim.tbl_map(function(client)
    return ('%d (%s)'):format(client.id, client.name)
  end, require('lspconfig.util').get_managed_clients())
end
```
and 

```lua
  desc = 'Manually launches a language server',
  nargs = '?',
  complete = function()
    return vim.tbl_keys(configs)
  end,
```

Comment (ii14) [reply plugin/lspconfig.lua:62]: Yes, you need to filter them then.

Comment (williamboman) [reply plugin/lspconfig.lua:62]: Ah, I wasn't aware you'd have to implement the filtering yourself, that explains it. Maybe this can be rectified in a later PR, this is starting to go a bit in circles with people chiming in haha

Comment (ranjithshegde) [reply plugin/lspconfig.lua:62]: Yes, getting this merged and then adding improvements if necessary maybe a good idea!

File: plugin/lspconfig.lua
Comment (ii14) [reply plugin/lspconfig.lua:1]: Disabling those commands and defining your own instead is a valid thing to do :)

File: lua/lspconfig/configs.lua
Comment (ranjithshegde) [reply lua/lspconfig/configs.lua:185]: Since `on_new_config` is called internally, didnt think it was necessary. But you are absolutely right, that was a mistake. Added it back

File: plugin/lspconfig.lua
Comment (ii14) [inline plugin/lspconfig.lua:54]:
```suggestion
    local config = require('lspconfig.configs')[server_name]
    if config then
      config.launch()
    end
```

**Comment by ranjithshegde** [comment]
Is this good to go then?

**Comment by ranjithshegde** [comment]
Is there a way to see on which file and line codespell action fails at?

**Comment by glepnir** [comment]
image

File: plugin/lspconfig.lua
Comment (justinmk) [inline plugin/lspconfig.lua:7]: has(nvim-0.7) will be true for all versions 7 or greater. It will only be false for 0.6 or older. The check should be has('nvim-0.8')

Comment (ranjithshegde) [reply plugin/lspconfig.lua:7]: I am a bit confused. The minimum required version is 0.7. it should fail if its lower than 0.7. but if this is set to 0.8, wont it fail for all versions other than nightly?

File: test/lspconfig_spec.lua
Comment (williamboman) [inline test/lspconfig_spec.lua:322]: This seems like a breaking change to me. `lspconfig.available_servers()` should still exist

Comment (ranjithshegde) [reply test/lspconfig_spec.lua:322]: Well at the moment, the `lspconfig.lua` simply sets up the metatable for `lspconfig.configs` and nothing else. If we add this, then it will setup a metatable + a single function that can actually be in utils.

Comment (williamboman) [reply test/lspconfig_spec.lua:322]: > Then the logic is simple. The portion of plugin simply sets up LSPs, with some commands in `plugin/` and some useful functions in `/utils`.

Yeah I agree it makes sense to remove it, I'm just thinking about not breaking parity with `master`. This will for sure break some user configs and plugins out there, although easily remedied.

> Also, too many PRs are being merged that potentially clash with this. Requires a bit of reshuffling

Yeah there seems to be a lot happening with this big PR still being open, I'll eject myself from further comments, let's get this merged haha

Comment (ranjithshegde) [reply test/lspconfig_spec.lua:322]: > Yeah I agree it makes sense to remove it, I'm just thinking about not breaking parity with `master`. This will for sure break some user configs and plugins out there, although easily remedied.
> 
Well this is changing/simplifying the core mechanism of the project. We could merge this as a breaking change, with a tag for previous versions?

File: plugin/lspconfig.lua
Comment (justinmk) [reply plugin/lspconfig.lua:7]: my mistake, I thought you wanted to check for 0.8

**Review by justinmk — APPROVED** [review]
Thanks everyone for the help here.

**Comment by ranjithshegde** [comment]
Squashed all the changes into single commit!

**Comment by justinmk** [comment]
Will merge this now. Can someone open a PR that _removes_ stuff for 0.8? It won't be merged until 0.8 is released, but a PR will help us figure out what changes are expected..

**Comment by ranjithshegde** [comment]
> Will merge this now. Can someone open a PR that _removes_ stuff for 0.8? It won't be merged until 0.8 is released, but a PR will help us figure out what changes are expected.

Will start with an issue. Maybe deprecation is requried as the first step. Many functions in `util` are available upstream.
A strategy could be to redirect those calls to upstream with a deprecation warning

**Comment by mike-lloyd03** [comment]
I almost just opened an issue for this but did a bit more digging first. I see that this commit introduces a breaking change. But I cannot tell from the conversation here what needs to be changed in my config to accommodate this change. Or if it's a problem with another plugin that depends on lspconfig and I should roll back to a previous version of this plug until other plug developers can update. Is there recommended guidance somewhere I should be looking at?

Thanks and sorry for this annoying message. I did my research first but came up short.

**Comment by justinmk** [comment]
@ranjithshegde  I wasn't expecting breaking changes here per https://github.com/neovim/nvim-lspconfig/pull/1984#issuecomment-1221625288  : 

> soft deprecate (show warning/error but continue to work) the commands API without a breaking change (so don't need to change the code)

What did we break?

**Comment by ii14** [comment]
@justinmk looks like `require("lspconfig").available_servers` is now `require("lspconfig.util").available_servers`.

**Comment by ii14** [comment]
If this is a problem, it could just be added back:
```diff
diff --git a/lua/lspconfig.lua b/lua/lspconfig.lua
index 1178202..c47707d 100644
--- a/lua/lspconfig.lua
+++ b/lua/lspconfig.lua
@@ -4,6 +4,10 @@ local M = {
   util = require 'lspconfig.util',
 }
 
+function M.available_servers()
+  return M.util.available_servers()
+end
+
 local mt = {}
 function mt:__index(k)
   if configs[k] == nil then
```

**Comment by glepnir** [comment]
@justinmk  like i said in #2078 I had search in github some configs and plugins use it .but it move to `utils.lua`

**Pull request merged**
